c7n/actions.py

```diff
@@ -822,3 +822,84 @@ def process_policy(self, policy, resource, matched_key):
         if not found:
             return None, found
         return statements, found
+
+
+class ModifyPolicyBase(BaseAction):
+
+    schema = utils.type_schema(
+        'modify-statements',
+        **{
+            'add-statements': {
+                'type': 'array',
+                'required': True,
+                'items': {
+                    'type': 'object',
+                    'properties': {
+                        'Sid': {'type': 'string'},
+                        'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
+                        'Principal': {'anyOf': [{'type': 'string'},
+                            {'type': 'object'}, {'type': 'array'}]},
+                        'NotPrincipal': {'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                        'Action': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'NotAction': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'Resource': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'NotResource': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'Condition': {'type': 'object'}
+                    },
+                    'required': ['Sid', 'Effect'],
+                    'oneOf': [
+                        {'required': ['Principal', 'Action', 'Resource']},
+                        {'required': ['NotPrincipal', 'Action', 'Resource']},
+                        {'required': ['Principal', 'NotAction', 'Resource']},
+                        {'required': ['NotPrincipal', 'NotAction', 'Resource']},
+                        {'required': ['Principal', 'Action', 'NotResource']},
+                        {'required': ['NotPrincipal', 'Action', 'NotResource']},
+                        {'required': ['Principal', 'NotAction', 'NotResource']},
+                        {'required': ['NotPrincipal', 'NotAction', 'NotResource']}
+                    ]
+                }
+            },
+            'remove-statements': {
+                'type': ['array', 'string'],
+                'oneOf': [
+                    {'enum': ['matched']},
+                    {'type': 'array', 'items': {'type': 'string'}},
+                    {'type': 'string'}
+                ],
+                'required': True
+            }
+        }
+    )
+
+    def add_policy(self, policy, resource):
+        current = {s['Sid']: s for s in policy.get('Statement', [])}
+        new = copy.deepcopy(current)
+        additional = {s['Sid']: s for s in self.data.get('add-statements', [])}
+        new.update(additional)
+        statements = list(new.values())
+        policy['Statement'] = statements
+        return policy
+
+    def remove_policy(self, policy, resource, matched_key):
```

Comment:
the method naming on the base class is a bit ambigious wrt to what its actually doing, ie its not really add_policy/remove_policy it add_statements/remove_statements.